examples/file-sharing.rs

```diff
@@ -0,0 +1,639 @@
+//!   network layer.
+//!
+//! - Single task driving the network layer, no locks required.
+//!
+//! ## Usage
+//!
+//! A two node setup with one node providing the file and one node requesting the file.
+//!
+//! 1. Run command below in one terminal.
+//!
+//!    ```
+//!    cargo run --example file-sharing -- \
+//!              --listen-address /ip4/127.0.0.1/tcp/40837 \
+//!              --secret-key-seed 1 \
+//!              provide \
+//!              --path <path-to-your-file> \
+//!              --name <name-for-others-to-find-your-file>
+//!    ```
+//!
+//! 2. Run command below in another terminal.
+//!
+//!    ```
+//!    cargo run --example file-sharing -- \
+//!              --peer /ip4/127.0.0.1/tcp/40837/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X \
+//!              get \
+//!              --name <name-for-others-to-find-your-file>
+//!    ```
+//!
+//! Note: The client does not need to be directly connected to the providing
+//! peer, as long as both are connected to some node on the same DHT.
+
+use async_std::io;
+use async_std::task::spawn;
+use futures::prelude::*;
+use libp2p::core::{Multiaddr, PeerId};
+use libp2p::multiaddr::Protocol;
+use std::error::Error;
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+#[async_std::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    env_logger::init();
+
+    let opt = Opt::from_args();
+
+    // Creates the network components, namely:
+    //
+    // - The network client to interact with the network layer from anywhere
+    //   within your application.
+    //
+    // - The network event stream, e.g. for incoming requests.
+    //
+    // - The network task driving the network itself.
+    let (mut network_client, mut network_events, network_task) =
+        network::new(opt.listen_address, opt.secret_key_seed)?;
+
+    // Spawn the network task for it to run in the background.
+    spawn(network_task);
+
+    // In case the user provided an address of a peer on the CLI, dial it.
+    if let Some(addr) = opt.peer {
+        let peer_id = match addr.iter().last() {
+            Some(Protocol::P2p(hash)) => PeerId::from_multihash(hash).expect("Valid hash."),
+            _ => return Err("Expect peer multiaddr to contain peer ID.".into()),
+        };
+        network_client.dial(peer_id, addr).await;
+    }
+
+    match opt.command {
+        // Providing a file.
+        Command::Provide { path, name } => {
+            // Advertise oneself as a provider of the file on the DHT.
+            network_client.start_providing(name.clone()).await;
+
+            loop {
+                match network_events.next().await {
+                    // Reply with the content of the file on incoming requests.
+                    Some(network::Event::InboundRequest { request, channel }) => {
+                        if request == name {
+                            let file_content = std::fs::read_to_string(&path)?;
```

Comment:
Use async fs module instead?

Reply:
Why? First I wouldn't expect an async version of `read_to_string` to lead to any performance gain, given that the actual context switching likely makes up only a small percentage of the overhead involved and second I don't think performance matters in the first place in this case.

Reply:
That depends a lot on how big the file is, no?

But in any case, the main argument I'd bring forward is that IMO examples shouldn't contain shortcuts that can present footguns but rather should show idiomatic code that can be taken and extended.

Using a sync fs API within an async fn can go wrong if the constraints or requirements around it change.

I don't feel super strongly about it though, so feel free to leave as is! :)